Fixes #37714 - inherit overrides deploy on in hostgroup

MariaAga opened the pull request:
if the hostgroup parent has a compute resource, the child hostgroup cant be set to have "bare metal", since empty value ("bare metal") = inherit.
In this pr I managed to make it work similarly to the host page, which is not perfect.
I added on top of `submit_with_all_params` to just reuse existing logic and not building new one.

sjha4 commented on app/views/hostgroups/_form.html.erb:
```diff
@@ -26,8 +26,16 @@
+          :id, 
+          :to_label,
+          { 
+            :disable_button_enabled => hostgroup_inherited_by_default?(:compute_resource_id, @hostgroup),
```
Do we need to hide this for parent hostgroups?

MariaAga replied:
No since they can also be children

adamruzicka replied:
What about for top-level host groups which don't have a parent?

MariaAga replied:
We could hide it, but it isn't hidden in hosts if they have no host group

MariaAga commented:
> I added on top of submit_with_all_params to just reuse existing logic and not building new one.

That didnt work, so I added an event listener

sjha4 commented:
The child hostgroup doesn't seem to be able to inherit the parent hostgroup. The compute resource gets set to empty/bare metal even when the form is set to inherit and populated with parent's compute resource.

sjha4 commented on app/assets/javascripts/host_edit.js:
```diff
@@ -165,6 +165,16 @@ function update_capabilities(capabilities) {
 var stop_pooling;
 
+if (window.location.href.contains('hostgroup')) {
+  document.addEventListener('DOMContentLoaded', function() {
```
Console error on this: `TypeError: window.location.href.contains is not a function`

MariaAga commented:
Changed the pr so now on submit hostgroup makes sure that the compute resource is included.
Also changed the inherit button behaviour for compute resource in hosts, when a user selects a host group the code saves the host groups compute resource, and when "inherit" is clicked it disables the select *and* selects the host groups compute resource. (in host & hostgroup forms)
before this pr in hosts if the compute resource was inherited it didnt always check the fields before the submit, so for example it didnt check libvirt connection if it was inherited, it does now

sjha4 commented:
Works now for setting compute resource id in the form. The inherit button is a bit confusing as when you click inherit on a child group, it populates the form as Bare Metal and disables the form which is not the inherited value..

MariaAga commented:
Thanks! added a fix that sets the inherit value on load, and not only on hostgroup change

sjha4 commented:
I have a parent HG for example with compute resource as libvirt-test:
![Screenshot from 2024-08-20 13-43-08](https://github.com/user-attachments/assets/7ffc4081-f54d-40b2-b33a-aaca74c670da)

On a child HG, the inherit button sets the compute resource as bare metal (nil) and then clicking the button switches enabled/disabled state of the input field. It's more logical to populate the field with parent's compute resource when the button is clicked vs setting it to bare metal?

![Screenshot from 2024-08-20 13-43-53](https://github.com/user-attachments/assets/7a66cb7e-f47a-4f36-9f80-97ffa5063b2f)
![Screenshot from 2024-08-20 13-44-10](https://github.com/user-attachments/assets/429133b1-8b36-4eaf-9722-64f23002fe33)

MariaAga commented:
Clicking on inherit should populate the deploy on with a the parent value,
screencap of parent libvirt, and parent bare metal
https://github.com/user-attachments/assets/b64398ab-6d6d-4661-a349-830a04da0f44

MariaAga commented:
removed an unused function

sjha4 commented:
Here's what I am seeing which is a little confusing:
When I click the nest button and land on a create HG page, it loads up with the correct compute resource selected and the field disabled.
![Nest-load](https://github.com/user-attachments/assets/1397755d-3e08-4995-add3-a413fc977b62)

If I click on "Inherit", the 1st click enables the Compute resource dropdown.
![Inherit-1stclick](https://github.com/user-attachments/assets/21e16cd6-8707-45b9-8a1a-ef8bc8d353b2)

If I click "Inherit" now, it sets compute resource to Bare metal and disables the dropdown.
Clicking inherit after this enables/disables the dropdown with Bare metal.
![Inherit-3rdClick](https://github.com/user-attachments/assets/7b8c937a-d1c5-45c5-9f5f-2a52f2670877)


 I'd expect the field to change to the actual inherited value: "test-compute-resource" whenever inherit button is clicked.

kmalyjur commented:
> If I click "Inherit" now, it sets compute resource to Bare metal and disables the dropdown.

It works fine for me - after clicking the button for the second time the compute resource stays there.

MariaAga commented:
Maybe emptying cache will resolve it? @sjha4

sjha4 commented:
I cleared my cache and opened a new incognito and this is what I see upon 2 clicks on inherit button on a nested create form.
[Screencast from 2024-08-28 16-23-42.webm](https://github.com/user-attachments/assets/9613bb69-fa28-43aa-8cc8-67df3025cc73)

MariaAga commented:
I managed to reproduce this error, but then I needed to restart the server (for something else) and the error was gone, so maybe thats the issue? sorry for this reproduction hunt @sjha4

MariaAga committed:
Fixes #37714 - inherit overrides deploy on in hostgroup

MariaAga commented:
Rebase fixed the failing tests

sjha4 reviewed (approved):
Ack 🎉

jeremylenz reviewed (approved):
ACK based on @sjha4 :+1:

dvo-rak commented:
Hello, 

I believe I am hitting a problem related to this MR after upgrading to 3.13 - once I edit a host and confirm, it tries to display the /hosts/xxx page (however the url still contains the /edit ...?), but the content is loading in an infinite loop.
![image](https://github.com/user-attachments/assets/cff4d031-fb8f-4970-8105-eca1bdedb8fc)

There is nothing in foreman/apache logs, but the browser console shows the following:
`Uncaught TypeError: document.querySelector(...) is null
    <anonymous> host_edit-a353532802cf651d1f8eaecab20ef1d6620e03c07c776b03559e9b18bebc15cf.js:1`

Any tips what I could be missing in my config/setup? I am not using any Compute Resources.

Thanks.

MariaAga commented:
@dvo-rak Thanks for the report, fixing it here: https://github.com/theforeman/foreman/pull/10447
